File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.rb
# @param {String} s
# @param {Integer} k
# @return {Integer}
def character_replacement(s, k)
    char_hash = Hash.new(0)
  res = 0
  left = 0

  for right in 0..(s.length-1)
    char_hash[s[right]] += 1

    while (right - left + 1) - char_hash.values.max > k
      char_hash[s[left]] -= 1
      left += 1
    end

    res = [res, (right - left + 1)].max
  end

  res
end

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var characterReplacement = function(s, k) {
    /*
        ABABCEF
        i=6,j=2,{A: 1, B: 1, C: 1, E: 1,F: 1}
            sum=0,max=0
    */
     let i=0,j=0,hash={}, result=0
    while(i<s.length){
        let sum=0, max=0
        if(hash[s[i]])
            hash[s[i]]+=1
        else
            hash[s[i]]=1
        Object.keys(hash).forEach(key=>{
            sum+=hash[key]
            if(hash[key]>max)
                max=hash[key]
        })
        
        if((sum-max)>k){
            hash[s[j]]-=1
            sum-=1
            j+=1
        }
        if(sum>result)
            result=sum
        i+=1
    }
    return result
};